src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java

```diff
@@ -579,28 +589,65 @@ private static String getStringInsideTag(String result, DetailNode detailNode) {
      * Finds and returns first sentence.
      *
      * @param ast Javadoc root node.
+     * @param period Period character.
      * @return first sentence.
      */
-    private static String getFirstSentence(DetailNode ast) {
+    private static String getFirstSentence(DetailNode ast, String period) {
+        final Deque<DetailNode> stack = new LinkedList<>();
+        stack.push(ast);
         final StringBuilder result = new StringBuilder(256);
-        final String periodSuffix = DEFAULT_PERIOD + ' ';
-        for (DetailNode child : ast.getChildren()) {
-            final String text;
-            if (child.getChildren().length == 0) {
-                text = child.getText();
+        boolean foundPeriod = false;
+        while (!stack.isEmpty()) {
+            final DetailNode node = stack.pop();
+            if (node.getChildren().length == 0
+                && appendUpToSentenceEndingPeriod(node.getText(), period, result)) {
+                foundPeriod = true;
+                break;
             }
-            else {
-                text = getFirstSentence(child);
+            // Pushing last child first means it will be processed last
+            for (int childIndex = node.getChildren().length - 1; childIndex >= 0; childIndex--) {
+                stack.push(node.getChildren()[childIndex]);
             }
+        }
+        if (!foundPeriod) {
+            result.setLength(0);
+        }
+        return result.toString();
+    }
 
-            if (text.contains(periodSuffix)) {
-                result.append(text, 0, text.indexOf(periodSuffix) + 1);
+    /**
+     * Find the end of a sentence, and append the sentence to the result if it ends with a period.
+     * If no period is present, append the whole string to the result. The end of sentence detection
+     * here could be replaced in the future by Java's built-in BreakIterator class.
+     *
+     * @param text string to append to result
+     * @param period period character to find
+     * @param result builder to append to
+     * @return true if a sentence ending period was found, false otherwise
+     */
+    private static boolean appendUpToSentenceEndingPeriod(
```

Comment:
Yes. I tried a lot of different ways to separate the methods. All the checks that need to pass can seem contradictory, and I pretty much gave up when I found a way to do it that the checks would accept. I can give it another go.

Would you suggest modelling the portion of the sentence that needs to be appended, and whether the period was found?

Reply:
I think I'll try and get rid of the StringBuilder. It's only there because it was there in the old version of the code. Most of the work that needs doing is combining the separate String objects from the different DetailNodes. There's only one case where we need to append only a part of the string, when we've found a period that ends the sentence. We don't need StringBuilder to do this efficiently.

Reply:
Sounds good, thanks for the quick reply.